admin/server-timing.php

```diff
@@ -0,0 +1,138 @@
+add_action( 'admin_menu', 'perflab_add_server_timing_page' );
+
+/**
+ * Initializes settings sections and fields for the Server-Timing page.
+ *
+ * @since n.e.x.t
+ */
+function perflab_load_server_timing_page() {
+	add_settings_section(
+		'benchmarking',
+		__( 'Benchmarking', 'performance-lab' ),
+		static function() {
+			?>
+			<p>
+				<?php esc_html_e( 'In this section, you can provide hook names to include measurements for them in the Server-Timing header.', 'performance-lab' ); ?>
+				<br>
+				<?php
+				echo wp_kses(
+					__( 'For any hook name provided, the <strong>cumulative duration between all callbacks</strong> attached to the hook is measured, in milliseconds.', 'performance-lab' ),
+					array( 'strong' => array() )
+				);
+				?>
+			</p>
+			<?php
+		},
+		PERFLAB_SERVER_TIMING_SCREEN
+	);
+
+	/*
+	 * For all settings fields, the field slug, option sub key, and label
+	 * suffix have to match for the rendering in the callback to be
+	 * semantically correct.
+	 */
+	add_settings_field(
+		'benchmarking_actions',
+		__( 'Actions', 'performance-lab' ),
+		static function() {
+			perflab_render_server_timing_page_field( 'benchmarking_actions' );
+		},
+		PERFLAB_SERVER_TIMING_SCREEN,
+		'benchmarking',
+		array( 'label_for' => 'server_timing_benchmarking_actions' )
+	);
+	add_settings_field(
+		'benchmarking_filters',
+		__( 'Filters', 'performance-lab' ),
+		static function() {
+			perflab_render_server_timing_page_field( 'benchmarking_filters' );
+		},
+		PERFLAB_SERVER_TIMING_SCREEN,
+		'benchmarking',
+		array( 'label_for' => 'server_timing_benchmarking_filters' )
+	);
+}
+
+/**
+ * Renders the Server-Timing page.
+ *
+ * @since n.e.x.t
+ */
+function perflab_render_server_timing_page() {
+	?>
+	<div class="wrap">
+		<?php settings_errors(); ?>
+		<h1>
+			<?php esc_html_e( 'Server-Timing', 'performance-lab' ); ?>
+		</h1>
+
+		<form action="options.php" method="post" novalidate="novalidate">
+			<?php settings_fields( PERFLAB_SERVER_TIMING_SCREEN ); ?>
+			<?php do_settings_sections( PERFLAB_SERVER_TIMING_SCREEN ); ?>
+			<?php submit_button(); ?>
+		</form>
+	</div>
+	<?php
+}
+
+/**
+ * Renders a field for the given Server-Timing option.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $slug Slug of the field and sub-key in the Server-Timing option.
+ */
+function perflab_render_server_timing_page_field( $slug ) {
+	$options = (array) get_option( PERFLAB_SERVER_TIMING_SETTING, array() );
+
+	// Value for the sub-key is an array of hook names.
+	$value = '';
+	if ( isset( $options[ $slug ] ) ) {
+		$value = implode( "\n", $options[ $slug ] );
+	}
+
+	$field_id       = "server_timing_{$slug}";
+	$field_name     = PERFLAB_SERVER_TIMING_SETTING . '[' . $slug . ']';
+	$description_id = "{$field_id}_description";
+
+	?>
+	<textarea
+		id="<?php echo esc_attr( $field_id ); ?>"
+		name="<?php echo esc_attr( $field_name ); ?>"
+		aria-describedby="<?php echo esc_attr( $description_id ); ?>"
+		class="large-text code"
+		rows="8"
```

Comment:
Without `novalidate`, custom validation could be added, for example:

```suggestion
		rows="8"
		onchange="this.setCustomValidity( /^\S+((\r?\n)+\S+)*\s*$/s.test( this.value ) ? '' : <?php echo esc_attr( wp_json_encode( __( 'Please enter one hook per line.', 'performance-lab' ) ) ); ?> )"
```

Nevertheless, as noted in the comment about `sanitize_key()`, I don't know actually if there is any validation that can be done since hook names can be basically anything.

Reply:
See my reply on that comment https://github.com/WordPress/performance/pull/784#discussion_r1286036347

Reply:
If the sanitization is removing all `[^A-Za-z0-9_\-\/\.]` characters, then adding this client-side validation would be important. Otherwise, the user could add such a custom hook and it would get silently munged when saving. Client-side validation would prevent that from happening. This should do it, along with removal of `novalidate`:

```suggestion
		rows="8"
		onchange="this.setCustomValidity( /^\s*([A-Za-z0-9_\-\/.]+\s*)*\s*$/s.test( this.value ) ? '' : <?php echo esc_attr( wp_json_encode( __( 'Please enter one hook per line and only use these characters: [A-Za-z0-9_\\-/\\.].', 'performance-lab' ) ) ); ?> )"
```

Reply:
Per my reply in https://github.com/WordPress/performance/pull/784#discussion_r1286036347, I'd prefer not to get hung up on that here, since it's not common in core either. There's tons of sanitization in core and little to no validation in most places.

Not saying this is something we can't improve, but client-side validation is by no means critical, the server-side sanitization however is. I'd prefer if we considered/discussed this separately.